File: sdk/typescript/src/providers/openai-compatible.ts
import type { ChatMessage, AnalysisResponse } from "../types.js";
import type {
  ProviderConfig,
  OpenAIRequestBody,
  OpenAIResponse,
  ResponseFormat,
} from "./types.js";

/**
 * OpenAI-compatible provider configuration
 * Uses OpenAI Chat Completions API format
 */
export const openaiCompatibleProvider: ProviderConfig = {
  baseUrl: process.env.OPENAI_COMPATIBLE_BASE_URL || "",
  envVar: "OPENAI_COMPATIBLE_API_KEY",

  authHeader: (apiKey: string) => ({
    Authorization: `Bearer ${apiKey}`,
    "Content-Type": "application/json",
  }),

  buildUrl: (baseUrl: string) => {
    if (!baseUrl) {
      throw new Error(
        "Missing OPENAI_COMPATIBLE_BASE_URL environment variable for openai-compatible provider"
      );
    }
    const trimmed = baseUrl.replace(/\/+$/, "");
    if (trimmed.endsWith("/chat/completions")) {
      return trimmed;
    }
    return `${trimmed}/chat/completions`;
  },

  transformRequest: (
    model: string,
    messages: ChatMessage[],
    responseFormat?: ResponseFormat
  ): OpenAIRequestBody => ({
    model,
    messages,
    temperature: 0,
    response_format: responseFormat,
  }),

  transformResponse: (response: unknown): AnalysisResponse => {
    const res = response as OpenAIResponse;
    return {
      id: res.id,
      usage: {
        promptTokens: res.usage.prompt_tokens,
        completionTokens: res.usage.completion_tokens,
        totalTokens: res.usage.total_tokens,
      },
      choices: res.choices.map((choice) => ({
        index: choice.index,
        message: {
          role: choice.message.role,
          content: choice.message.content,
        },
        finish_reason: choice.finish_reason,
      })),
    };
  },
};

File: sdk/python/src/safety_agent/providers/openai_compatible.py
"""
OpenAI-compatible provider configuration (Chat Completions API)
"""

import os
from typing import Any

from ..types import ChatMessage, AnalysisResponse, TokenUsage, AnalysisResponseChoice
from ..schemas import ResponseFormat


class OpenAICompatibleProvider:
    """OpenAI-compatible provider configuration using Chat Completions API."""

    base_url = os.environ.get("OPENAI_COMPATIBLE_BASE_URL", "")
    env_var = "OPENAI_COMPATIBLE_API_KEY"

    def auth_header(self, api_key: str) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        }

    def build_url(self, base_url: str, model: str) -> str:
        if not base_url:
            raise ValueError(
                "Missing OPENAI_COMPATIBLE_BASE_URL environment variable for openai-compatible provider"
            )
        trimmed = base_url.rstrip("/")
        if trimmed.endswith("/chat/completions"):
            return trimmed
        return f"{trimmed}/chat/completions"

    def transform_request(
        self,
        model: str,
        messages: list[ChatMessage],
        response_format: ResponseFormat | None = None,
    ) -> dict[str, Any]:
        request: dict[str, Any] = {
            "model": model,
            "messages": [
                {
                    "role": msg.role,
                    "content": msg.content if isinstance(msg.content, str) else msg.content,
                }
                for msg in messages
            ],
        }

        if response_format:
            request["response_format"] = response_format

        return request

    def transform_response(self, response: dict[str, Any]) -> AnalysisResponse:
        choices = response.get("choices", [])
        content = ""
        if choices:
            message = choices[0].get("message", {})
            content = message.get("content", "")

        usage = response.get("usage", {})
        return AnalysisResponse(
            id=response.get("id", ""),
            usage=TokenUsage(
                prompt_tokens=usage.get("prompt_tokens", 0),
                completion_tokens=usage.get("completion_tokens", 0),
                total_tokens=usage.get("total_tokens", 0),
            ),
            choices=[
                AnalysisResponseChoice(
                    index=0,
                    message=ChatMessage(role="assistant", content=content),
                    finish_reason=choices[0].get("finish_reason", "stop")
                    if choices
                    else "stop",
                )
            ],
        )


openai_compatible_provider = OpenAICompatibleProvider()
